File: backend/schemas/reports.py
"""
Report schemas for student performance and progress reports.
"""

from datetime import date
from enum import Enum
from typing import List, Optional

from pydantic import BaseModel, ConfigDict, Field


class ReportFormat(str, Enum):
    """Supported report output formats."""

    JSON = "json"
    PDF = "pdf"
    CSV = "csv"


class ReportPeriod(str, Enum):
    """Report time period options."""

    WEEK = "week"
    MONTH = "month"
    SEMESTER = "semester"
    YEAR = "year"
    CUSTOM = "custom"


class PerformanceReportRequest(BaseModel):
    """Request schema for generating student performance report."""

    student_id: int = Field(..., description="Student ID")
    course_ids: Optional[List[int]] = Field(
        None,
        description="Optional list of course IDs to include (all courses if empty)",
    )
    start_date: Optional[date] = Field(
        None, description="Report start date (for custom period)"
    )
    end_date: Optional[date] = Field(
        None, description="Report end date (for custom period)"
    )
    period: ReportPeriod = Field(
        ReportPeriod.SEMESTER, description="Report time period"
    )
    format: ReportFormat = Field(ReportFormat.JSON, description="Output format")
    include_grades: bool = Field(True, description="Include grade information")
    include_attendance: bool = Field(True, description="Include attendance information")
    include_daily_performance: bool = Field(
        True, description="Include daily performance scores"
    )
    include_highlights: bool = Field(True, description="Include semester highlights")


class PerformanceSummary(BaseModel):
    """Summary statistics for a performance category."""

    category: str
    total_entries: int
    average_score: float
    min_score: float
    max_score: float
    average_percentage: float
    trend: str  # "improving", "declining", "stable"


class CourseSummary(BaseModel):
    """Course-level summary in performance report."""

    course_id: int
    course_code: str
    course_name: str
    grade_average: Optional[float] = None
    grade_percentage: Optional[float] = None
    attendance_rate: Optional[float] = None
    total_absences: Optional[int] = None
    performance_categories: List[PerformanceSummary] = []
    latest_grade: Optional[float] = None
    weighted_grade: Optional[float] = None


class AttendanceSummary(BaseModel):
    """Attendance summary statistics."""

    total_days: int
    present: int
    absent: int
    late: int
    excused: int
    attendance_rate: float
    unexcused_absences: int


class GradeSummary(BaseModel):
    """Grade summary statistics."""

    total_assignments: int
    average_grade: float
    average_percentage: float
    highest_grade: float
    lowest_grade: float
    grade_trend: str  # "improving", "declining", "stable"


class HighlightSummary(BaseModel):
    """Highlight information for the period."""

    semester: str
    rating: Optional[int]
    category: str
    text: str
    is_positive: bool
    date_created: date


class StudentPerformanceReport(BaseModel):
    """Complete student performance report response."""

    model_config = ConfigDict(
        json_schema_extra={
            "example": {
                "student_id": 1,
                "student_name": "John Doe",
                "student_email": "john@example.com",
                "report_period": "semester",
                "start_date": "2025-09-01",
                "end_date": "2025-12-31",
                "generated_at": "2025-12-12T10:30:00Z",
                "overall_attendance": {
                    "total_days": 90,
                    "present": 85,
                    "absent": 5,
                    "attendance_rate": 94.4,
                },
                "courses": [
                    {
                        "course_id": 1,
                        "course_code": "CS101",
                        "course_name": "Introduction to Programming",
                        "grade_average": 85.5,
                        "attendance_rate": 95.0,
                    }
                ],
            }
        }
    )

    student_id: int
    student_name: str
    student_email: str
    report_period: str
    start_date: date
    end_date: date
    generated_at: str

    # Overall summaries
    overall_attendance: Optional[AttendanceSummary] = None
    overall_grades: Optional[GradeSummary] = None

    # Course-level breakdown
    courses: List[CourseSummary] = []

    # Highlights
    highlights: List[HighlightSummary] = []

    # Recommendations
    recommendations: List[str] = []


class BulkReportRequest(BaseModel):
    """Request schema for generating reports for multiple students."""

    student_ids: List[int] = Field(..., description="List of student IDs")
    course_ids: Optional[List[int]] = Field(None, description="Optional course filter")
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    period: ReportPeriod = ReportPeriod.SEMESTER
    format: ReportFormat = ReportFormat.JSON
    include_grades: bool = True
    include_attendance: bool = True
    include_daily_performance: bool = True
    include_highlights: bool = True


class ReportJobStatus(BaseModel):
    """Status of an asynchronous report generation job."""

    job_id: str
    status: str  # "pending", "processing", "completed", "failed"
    progress: int  # 0-100
    created_at: str
    completed_at: Optional[str] = None
    download_url: Optional[str] = None
    error_message: Optional[str] = None

File: frontend/src/locales/el/reports.js
// Greek translations for Reports module (Ελληνικές μεταφράσεις για τη λειτουργία Αναφορών)
export default {
  // Report types
  studentPerformanceReport: 'Αναφορά Απόδοσης Μαθητή',
  bulkReport: 'Μαζική Αναφορά',

  // Actions
  generate: 'Δημιουργία Αναφοράς',
  generating: 'Δημιουργία Αναφοράς...',
  newReport: 'Νέα Αναφορά',
  print: 'Εκτύπωση Αναφοράς',
  download: 'Λήψη',
  export: 'Εξαγωγή',

  // Configuration
  configuration: 'Ρυθμίσεις',
  period: 'Περίοδος Αναφοράς',
  lastWeek: 'Τελευταία Εβδομάδα',
  lastMonth: 'Τελευταίος Μήνας',
  semester: 'Εξάμηνο',
  year: 'Ακαδημαϊκό Έτος',
  custom: 'Προσαρμοσμένη Περίοδος',
  period_week: 'Εβδομάδα',
  period_month: 'Μήνας',
  period_semester: 'Εξάμηνο',
  period_year: 'Έτος',
  period_custom: 'Προσαρμοσμένο',

  // Include options
  includeData: 'Συμπερίληψη Δεδομένων',
  includeGrades: 'Συμπερίληψη Βαθμών',
  includeAttendance: 'Συμπερίληψη Παρουσιών',
  includePerformance: 'Συμπερίληψη Απόδοσης',
  includeHighlights: 'Συμπερίληψη Επισημάνσεων',

  // Summaries
  attendanceSummary: 'Σύνοψη Παρουσιών',
  gradesSummary: 'Σύνοψη Βαθμών',
  performanceSummary: 'Σύνοψη Απόδοσης',
  courseBreakdown: 'Ανάλυση ανά Μάθημα',
  recommendations: 'Συστάσεις',
  highlights: 'Επισημάνσεις',

  // Attendance metrics
  attendanceRate: 'Ποσοστό Παρουσιών',
  totalDays: 'Συνολικές Ημέρες',
  present: 'Παρών',
  absent: 'Απών',
  late: 'Καθυστέρηση',
  excused: 'Δικαιολογημένος',
  unexcusedAbsences: 'Αδικαιολόγητες Απουσίες',
  attendanceRateValue: '{{rate}}% ({{present}}/{{total}} ημέρες)',

  // Grade metrics
  average: 'Μέσος Όρος',
  totalAssignments: 'Συνολικές Εργασίες',
  highest: 'Υψηλότερος',
  lowest: 'Χαμηλότερος',
  trend: 'Τάση',
  grade: 'Βαθμός',
  attendance: 'Παρουσία',
  absences: 'Απουσίες',

  // Trends
  trend_improving: 'Βελτιούμενη',
  trend_declining: 'Φθίνουσα',
  trend_stable: 'Σταθερή',

  // Performance
  performanceCategories: 'Απόδοση ανά Κατηγορία',
  category: 'Κατηγορία',
  averageScore: 'Μέση Βαθμολογία',

  // Recommendations (Συστάσεις)
  rec_attendance_low: '⚠️ Η παρουσία είναι κάτω από 75%. Η τακτική παρουσία είναι κρίσιμη για την ακαδημαϊκή επιτυχία.',
  rec_attendance_medium: '✓ Η παρουσία θα μπορούσε να βελτιωθεί. Στοχεύστε για ποσοστό παρουσίας 90%+.',
  rec_grades_failing: '⚠️ Οι βαθμοί είναι κάτω από το όριο επιτυχίας. Εξετάστε το ενδεχόμενο προγραμματισμού ιδιαίτερων μαθημάτων.',
  rec_grades_needs_improvement: '✓ Οι βαθμοί χρειάζονται βελτίωση. Επανεξετάστε τις μεθόδους μελέτης και ζητήστε βοήθεια στα δύσκολα μαθήματα.',
  rec_grades_excellent: '🌟 Εξαιρετική ακαδημαϊκή επίδοση! Συνεχίστε την εξαιρετική δουλειά.',
  rec_trend_declining: '⚠️ Η τάση των βαθμών είναι φθίνουσα. Επανεξετάστε τις πρόσφατες εργασίες και εντοπίστε τομείς που χρειάζονται βελτίωση.',
  rec_trend_improving: '✓ Η τάση των βαθμών βελτιώνεται. Συνεχίστε την θετική δυναμική!',
  rec_focus_courses: '📚 Εστιάστε σε: {{courses}}. Εξετάστε επιπλέον χρόνο μελέτης ή ιδιαίτερα μαθήματα.',
  rec_satisfactory: '✓ Η συνολική επίδοση είναι ικανοποιητική. Συνεχίστε να διατηρείτε καλές συνήθειες μελέτης.',

  // Messages
  error: 'Σφάλμα κατά τη δημιουργία αναφοράς. Παρακαλώ προσπαθήστε ξανά.',
  noData: 'Δεν υπάρχουν διαθέσιμα δεδομένα για την επιλεγμένη περίοδο.',
  reportGenerated: 'Η αναφορά δημιουργήθηκε επιτυχώς',

  // Status
  pending: 'Εκκρεμεί',
  processing: 'Επεξεργασία',
  completed: 'Ολοκληρώθηκε',
  failed: 'Απέτυχε'
};
